File: binary_tree/preorder_traversal.rb
# 144
# a preorder traversal whose basic operation is pushing a node's value into an array
def preorder_traversal(root, vals=[])
  # recursive base case
  return [] if root.nil?

  # perform the operation on the current node
  vals << root.val

  # recursively perform operation on left child
  preorder_traversal(root.left, vals)

  # recursively perform operation on left child
  preorder_traversal(root.right, vals)

  # to satisfy requirement that initial invocation return array of vals in preorder order
  vals
end

File: binary_tree/preorder_traversal.js
// https://leetcode.com/problems/binary-tree-postorder-traversal/description/
// given a binary tree, return the postorder traversal of its nodes' values.
var postorderTraversal = function(root) {
  const vals = [];
  helper(root, vals);
  return vals;
};

function helper(node, vals) {
  if (node === null) {
    return null;
  }

  helper(node.left, vals);
  helper(node.right, vals);
  vals.push(node.val);
}
